Add support and test for multi-D data

RyanJHeld opened the pull request:
I think the code looks pretty ugly, but we talked about cleaning this up in the next ticket (LATIS-847). It works with the given 2-D test, and should work work for even more dimensions. I would like to add a test for a higher dimension dataset, but I'm trying to get this PR in before sprint close by the skin of my teeth.

I split the `Accumulate` function into two parts. The first is still `Accumulate`, the second is `domainAccumulatorToNcArray` for domain variables, and `AccumulatorToNcArray` for the range. The point of this is so duplicate values in the domain can be removed before converting them to NetCDF Arrays and written to  file.

Commit:
Add support and test for multi-D data

RyanJHeld commented:
I'm going to add a few assumptions to the scaladoc.

Commit:
Add assumption to scaladoc

remove unused private method
remove unnecessary semicolons

sapols commented on netcdf/src/main/scala/latis/output/NetcdfEncoder.scala:
```diff
@@ -5,8 +5,8 @@ import java.io.File
 import fs2.Stream
-import ucar.ma2.{DataType => NcDataType}
 import ucar.ma2.{Array => NcArray}
+import ucar.ma2.{DataType => NcDataType}
```
I changed the imports in my last PR to not group them with curlies. Let’s standardize on that.

lindholc replied:
This is different. This is renaming the thing being imported.

sapols replied:
Ah, yes.

Commit:
Add 3-D test for NetCDF encoder

Change 2-D test dataset to be curried to test the uncurrying in the encoder
Remove wrong assumption from NetCDF encoder description

lindholc commented on netcdf/src/main/scala/latis/output/NetcdfEncoder.scala:
```diff
@@ -113,12 +125,57 @@ object NetcdfEncoder {
           throw LatisException(
-            "dataset must be a function of a scalar to either a scalar or tuple of scalars"
+            "dataset must be a function where domain and range are scalars or tuples of scalars"
           )
```
Can we even reach this case? If we're iterating through a `Seq[Sample]` and we aren't constraining `d` or `r`, it seems like the first pattern should always match.

RyanJHeld replied:
Good point. What should I put in place of the exception? Will I get a compiler warning if I leave off the `case _`?

lindholc replied:
I think we just remove it entirely.

I didn't get a warning when I removed the `_` case. That doesn't always mean it's safe, but I think it is in this case. We know what we're matching against is a `Sample`, which is just a certain kind of tuple. This match is less about choosing between alternatives and more about binding `d` and `r`.

Like this:

```scala
val x: (Int, Int) = ???

x match {
  case (a, b) => ???
  // What other cases are there?
}
```

lindholc replied:
I think `Sample(d, r)` is an [irrefutable constructor pattern](https://www.scala-lang.org/files/archive/spec/2.12/08-pattern-matching.html#irrefutable-patterns) (case 3).

lindholc reviewed (approved):
I'll wait for the other PR for more detailed review.

I did leave a comment on something because I thought about it now and I'll probably forget before the next PR.

dlindhol commented on netcdf/src/test/scala/latis/output/NetcdfEncoderSpec.scala:
```diff
@@ -106,4 +155,56 @@ object NetcdfEncoderSpec {
+    )
+
+    val md = Metadata("time_series_2D")
+    val model = Function(
```
I created `latis.model.ModelParser` so you can do things like:
```
val model = ModelParser.parse("time -> wavelength -> (flag, flux)")
```
It's handy if you don't care about the types. It expects every scalar to be a double.

RyanJHeld replied:
Oh that's handy. Can you define the types after, or are they stuck as Double?

RyanJHeld replied:
I think it would be cool if you could define a model like 
```
val model = ModelParser.parse("time: int -> wavelength: double -> (flag: byte, flux: double)")
```

sapols replied:
That would be cool.

dlindhol replied:
I have been thinking about a richer form of the model syntax like that. (There is a TODO in the code for this.) Consider named tuples, for example:
`myVector: (x, y, z)`
where the type is indicated with the "()". It seems consistent to add types to the scalars with ":":
`myVector: (x: int, y: int, z: int)`
Since we were talking about Cartesian datasets before, I've though about square brackets to indicate a Cartesian domain set: `[x, y] -> f` as opposed to any set of xs and ys: `(x, y) -> f`.

dlindhol commented on netcdf/src/test/scala/latis/output/NetcdfEncoderSpec.scala:
```diff
@@ -106,4 +155,56 @@ object NetcdfEncoderSpec {
     new MemoizedDataset(md, model, data)
   }
+
+  private val time_series_2D: MemoizedDataset = {
```
See `latis.util.DatasetGenerator` in latis3 test for ideas to make dataset generation more convenient for testing (and the DSL?).

dlindhol commented on netcdf/src/test/scala/latis/output/NetcdfEncoderSpec.scala:
```diff
@@ -106,4 +155,56 @@ object NetcdfEncoderSpec {
+    )
+    val data = SampledFunction(samples)
+
+    new MemoizedDataset(md, model, data)
```
I'm surprised if something doesn't go wrong here. Since the model specifies a nested function: `time -> wavelength -> (flag, flux)`, each sample should have a SampledFunction in the range. Your sample definitions would call for the model: `(time, wavelength) -> (flag, flux)`.

RyanJHeld replied:
So something like `Sample(DomainData(1, 430.1), RangeData(0: Byte, 1.0))` would be 
```
Sample(
  DomainData(1), 
  SampledFunction(
    Sample(
      DomainData(430.1), 
      RangeData(0: Byte, 1.0)
    )
  )
)
```
?

RyanJHeld replied:
I think I figured it out. I didn't quite have it above, but I got it now.

dlindhol replied:
For each outer Sample, you would have to construct a SampledFunction for the range. It's a lot easier to define the uncurried form then call `curry` on it. Or maybe do the brute force thing in the reusable DatasetGenerator.

dlindhol commented on netcdf/src/test/scala/latis/output/NetcdfEncoderSpec.scala:
```diff
@@ -64,6 +64,55 @@ class NetcdfEncoderSpec extends FlatSpec {
+
+    val file   = enc.encode(time_series_2D).compile.toList.unsafeRunSync().head
+    val ncFile = NetcdfFile.open(file.getAbsolutePath)
+    try {
```
I can imagine wrapping this as a managed resource instead of the try/finally. Or just map over the Stream that we get from the encoder?

RyanJHeld replied:
I'll make note of this for the refactor.

lindholc replied:
It would also be better to use a temporary file rather than make a file in the project root.

dlindhol commented on netcdf/src/main/scala/latis/output/NetcdfEncoder.scala:
```diff
@@ -16,15 +16,16 @@ import latis.data.Data
  *
- * This encoder assumes the dataset is uncurried.
+ * This encoder assumes:
+ *   - the dataset is sorted so that the first domain variable changes slowest
```
I think the ultimate assumption here is that we have a Cartesian Dataset: the domain is a product of  the values for each dimension. We are effectively undoing that product here to get the separate domain variables. The ordering is as you say, but I think that is a LaTiS level definition of how we implement ordering for a Cartesian Dataset. One of these days I hope to have something like a `CatesianDataset` trait that we can mix in so we can let the types express the limitations.

RyanJHeld replied:
Yes, I should add that a cartesian dataset is assumed, thanks for pointing that out. The ordering is not enforced anywhere, right? If that's the case, I'll leave that in the assumption.

dlindhol replied:
We can, in theory, specify any Ordering. See `Scalar.ordering` which is intended to provide the appropriate ordering for any given scalar based on its metadata. If the samples don't obey expected ordering, then there is a "bug" in the dataset. That makes the assumption become "there are no bugs in the dataset" which could be left unspoken.

RyanJHeld replied:
I took a look at `Scalar.ordering` and I think I get it. It makes sense to leave off this assumption, but I don't actually check the ordering in the encoder. It looks like the ordering can only be ascending or descending. Shouldn't I be checking to make sure the ordering isn't descending?

dlindhol replied:
NetCDF doesn't have any ordering requirements. We should be able to preserve the ordering no matter what it is. I think we only need a sense of equality for "distinct" domain values. Not sure how that is implemented. Note that scala has the `Equiv` type class (or trait, anyway) for things that also have an `Ordering`.  Support for ordering is still half-baked in latis3 but it's more that we had in v2.

dlindhol reviewed (commented):
Submitting one round of comments. I think the 2D test dataset model/sample inconsistency is the only thing that needs fixing, so far.

dlindhol commented on netcdf/src/main/scala/latis/output/NetcdfEncoder.scala:
```diff
@@ -61,20 +62,29 @@ class NetcdfEncoder(file: File) extends Encoder[IO, File] {
     model: DataType,
     datasetList: Vector[Sample]
```
I would generally use `samples` for a variable like this. It might also make more sense to name the method `samplesToNetcdf`. However, it seems more natural to me to pass the Dataset as an argument, then actually stream the samples to the accumulator. This would presumably reduce memory use. And presumably we could use NetCDF chunking (eventually).

RyanJHeld replied:
I'll put this in the refactor.

dlindhol commented on netcdf/src/main/scala/latis/output/NetcdfEncoder.scala:
```diff
@@ -61,20 +62,29 @@ class NetcdfEncoder(file: File) extends Encoder[IO, File] {
     model: DataType,
     datasetList: Vector[Sample]
   ): Unit = model match {
```
Smells like a case for `Either`. Presumably there are a number of exceptions that netcdf might throw that we might want to wrap.

lindholc replied:
I think it ought to be `IO[Unit]`. `Either` might make sense if this were pure, but I'd assume any method called on `file` will be impure.

RyanJHeld replied:
Wouldn't wrapping the method calls on `file` in `IO` make this function pure?

I tried to simply change the return type to `IO[Unit]` and wrap the last two lines in `IO`, and while that does compile, the tests don't pass. I think I will put this in the refactor.

lindholc replied:
Yes, if you were to wrap all the calls on `file` so they return `IO`, you'd be forced to write this method such that the whole thing returns `IO` (assuming you don't use an `unsafeRun` method).

RyanJHeld replied:
Would it have to be every call to `file`? I think all the `addDimension` and `addVariable` calls happen in memory, and the file doesn't get created until `file.create()`. If the other calls to `file` are in fact pure, then it's just the last two lines that would be wrapped in `IO`.

lindholc replied:
Remember that `IO` isn't about file IO. It's about preserving referential transparency (and other things).

Do these snippets do the same thing?

```scala
// Snippet 1
val x: Variable = file.addVariable(...)
(x, x)

// Snippet 2
(file.addVariable(...), file.addVariable(...))
```

Snippet 1 adds a single variable to the file and returns it paired with itself. Snippet 2 adds two variables to the file and returns them in a pair. We know `file.addVariable` isn't referentially transparent and needs to be wrapped in `IO` because the snippets do not do the same thing.

Commit:
Fix things for PR

dlindhol commented on netcdf/src/main/scala/latis/output/NetcdfEncoder.scala:
```diff
@@ -162,13 +218,4 @@ object NetcdfEncoder {
       // Boolean is not supported by netCDF3
       case t => throw LatisException(s"Unsupported type: $t")
```
This is making me think of a validation pattern (design by contract?) that would be performed eagerly. Upon calling any operation or encoder with a Dataset, it would verify that the Dataset meets all the assumptions. It would also provide a clearer picture of what the assumptions are. We could have cheap metadata/model level validation (e.g. supported types) and expensive data level validation (e.g. all samples are ordered).

dlindhol commented on netcdf/src/main/scala/latis/output/NetcdfEncoder.scala:
```diff
@@ -61,20 +63,29 @@ class NetcdfEncoder(file: File) extends Encoder[IO, File] {
+      val rScalars = range.getScalars
+      val scalars  = dScalars ++ rScalars
+      val acc: Accumulator =
+        accumulate(scalarsToAccumulator(scalars, datasetList.length), scalars, datasetList)
```
It seems to me that there is no need to worry about types, yet. Could we just accumulate the `Anys` then do the type matching when adding them to NetCDF?
Another thought is whether we could use `ma2.Arrays` as the accumulators in the first place instead of the intermediate scala Arrays.

RyanJHeld replied:
I'm not sure how to get the value out of a `Data` without knowing the type. For example, `v.asInstanceOf[ByteValue].value` works, but `v.value` and `v.asInstanceOf[Any].value` does not.

I might be able to put things into `ma2.Arrays` directly. Working with them is not easy since they don't support a lot of the functionality Scala arrays do, but it might be possible. I use the `distinct` method on array, and `ma2.Arrays` doesn't have an equivalent (that I know of), so I would have to do that myself.

dlindhol replied:
`Datum` has a `value` method that returns `Any`. If you have `Scalar` types then you should have `Datum`. Since we uncurried, we should have nothing but scalar `Datum` in the `Samples`.

RyanJHeld replied:
I tried replacing the pattern matching of `s.valueType` with 
```
case (v, s, a) => a.asInstanceOf[Array[Any]](i) = v.asInstanceOf[Datum].value
```
but I get 
```
java.lang.ClassCastException: [I cannot be cast to [Ljava.lang.Object;
```
Am I doing it wrong?

dlindhol replied:
It looks like you are running into the invariance of Arrays. Some other `Seq` type might work. I also wondered about using something like `ArrayBuffer` for accumulation so you don't have to know the size. There is cost in re-allocation, though.

Another thought: can you use a pattern match instead of casting with `asInstanceOf`?

RyanJHeld replied:
There doesn't seem to be a trivial way to do this (that I know of). I can explore this further if you would like, but I would want to know that that's actually something you want me to do. If so, I would prefer to put that in the next ticket.

About pattern matching, is there are particular use of `asInstanceOf` you're looking at, or are you thinking of replacing all of them? I'm not sure how I could make that work for either `accumulate` or `domainAccumulatorToNcArray`, but if you would like me to look into it, I could do that for the refactor.

dlindhol replied:
I'm fine with saving that for later. I'm curious if we can come up with a generic "columnizing" technique that we could reuse in places like this.

dlindhol commented on netcdf/src/main/scala/latis/output/NetcdfEncoder.scala:
```diff
@@ -61,20 +63,29 @@ class NetcdfEncoder(file: File) extends Encoder[IO, File] {
 
+      val dArrs = domainAccumulatorToNcArray(acc, dScalars)
+      val shape = dArrs.map(_.getSize.toInt).toArray
+      val rArrs = accumulatorToNcArray(acc.slice(dScalars.length, acc.length), rScalars, shape)
```
How about `acc.drop(dScalars.length)` instead of the slice?

RyanJHeld replied:
`drop` does seem better. I'll do that.

Commit:
Minor fixes

Commit:
Make pattern match consistent
